### PR TITLE
[SYCL] Add property set types and JSON representation

### PR DESCRIPTION
**Pull request by jzc**
This PR adds the `PropertySet` type, along with a pair of functions used to serialize and deserialize into a JSON representation. A property set is a key-value map, with values being one of 2 types - uint32 or byte array. A property set registry is a collection of property sets, indexed by a "category" name.

In SYCL offloading, property sets will be used to communicate metadata about device images needed by the SYCL runtime. For example, there is a property set which has a byte array containing the numeric ID, offset, and size of each SYCL2020 spec constant. Another example is a property set describing the optional kernel features used in the module: does it use fp64? fp16? atomic64? 

This metadata will be computed by `clang-sycl-linker` and the JSON representation will be inserted in the string table of each output `OffloadBinary`.  This JSON will be consumed the SYCL offload wrapper and will be lowered to the binary form SYCL runtime expects. 

For example, consider this SYCL program that calls a kernel that uses fp64:

```c++
#include <sycl/sycl.hpp>

using namespace sycl;
class MyKernel;

int main() {
  queue q;
  auto *p = malloc_shared<double>(1, q);
  *p = .1;
  q.single_task<MyKernel>([=]{ *p *= 2; }).wait();
  std::cout << *p << "\n";
  free(p, q);
}
```

The device code for this program would have the kernel marked with `!sycl_used_aspects`:

```
define spir_kernel void @_ZTS8MyKernel([...]) !sycl_used_aspects !n { [...] }
!n = {i32 6}
```

`clang-sycl-linker` would recognize this metadata and then would output the following JSON in the `OffloadBinary`'s key-value map:

```
{
  "SYCL/device requirements": {
    // aspects contains a list of sycl::aspect values used
    // by the module; in this case just the value 6 encoded 
    // as a 4-byte little-endian integer 
    "aspects": "BjAwMA=="
  }
}
```

The SYCL offload wrapper would lower those property sets to something like this:

```
struct _sycl_device_binary_property_set_struct {
  char *CategoryName;
  _sycl_device_binary_property *PropertiesBegin;
  _sycl_device_binary_property *PropertiesEnd;
};

struct _sycl_device_binary_property_struct {
  char *PropertyName;  
  void *ValAddr;     
  uint64_t ValSize; 
};

//  
_sycl_device_binary_property_struct device_requirements[] = {
  /* PropertyName */ "aspects",
  /* ValAddr */ [pointer to the bytes 0x06 0x00 0x00 0x00],
  /* ValSize */ 4,
};

_sycl_device_binary_property_set_struct properties[] = {
  /* CategoryName */ "SYCL/device requirements",
  /* PropertiesBegin */ device_requirements,
  /* PropertiesEnd */ std::end(device_requirments),
}
```

### CODE REVIEW

File: llvm/include/llvm/Frontend/Offloading/Utility.h [hunk 1]
```diff
@@ -159,6 +160,107 @@ namespace intel {
...
+class PropertySetRegistry;
+
+// A property value. It can be either a 32-bit unsigned integer or a byte array.
+class PropertyValue {
+public:
+  using ByteArrayTy = SmallVector<char, 8>;
+
+  PropertyValue() = default;
+  PropertyValue(uint32_t Val) : Value(Val) {}
+  PropertyValue(StringRef Data)
```
Comment (jzc) [inline llvm/include/llvm/Frontend/Offloading/Utility.h]: I've simplified the byte array code compared to the downstream code, but they won't compatible downstream and would cause ABI-break and runtime changes. The downstream implementation always wrote the number of bits the data uses in the first 8 bytes of a byte-array property value, but after some tests, I found this values way never used, except for one place that could be worked around.

[https://github.com/intel/llvm/blob/a9db58476ab1e1ff87f128088ac203539d87d22b/llvm/lib/Support/PropertySetIO.cpp#L153-L169](https://github.com/intel/llvm/blob/a9db58476ab1e1ff87f128088ac203539d87d22b/llvm/lib/Support/PropertySetIO.cpp#L153-L169)

File: llvm/include/llvm/Frontend/Offloading/Utility.h
Comment (jzc) [inline llvm/include/llvm/Frontend/Offloading/Utility.h:1]: For now, I've placed the class in `Frontend/Offloading/Utility.h`. If we get `Frontend/Offloading/SYCL` later on then I then that would be the most appropriate place to put it.

Comment (asudarsa) [reply llvm/include/llvm/Frontend/Offloading/Utility.h:1]: I think it might be a better idea to add this in a separate file called PropertySet.h (associated impl in PropertySet.cpp).

Thanks

Comment (asudarsa) [reply llvm/include/llvm/Frontend/Offloading/Utility.h:1]: Adding @sarnex for comment as he has worked with Utility.h before.

Thanks

File: llvm/include/llvm/Frontend/Offloading/Utility.h [hunk 2]
```diff
@@ -159,6 +160,107 @@ namespace intel {
 /// the Intel runtime offload plugin.
 Error containerizeOpenMPSPIRVImage(std::unique_ptr<MemoryBuffer> &Binary);
 } // namespace intel
+
+namespace sycl {
+class PropertySetRegistry;
+
```
Comment (asudarsa) [inline llvm/include/llvm/Frontend/Offloading/Utility.h]: Why do we need this?

Thanks

File: clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp [hunk 3]
```diff
@@ -20,6 +20,7 @@
 #include "llvm/BinaryFormat/Magic.h"
```
Comment (asudarsa) [inline clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp]: Changes to ClangSYCLLinker.cpp can be done in a separate PR. WDYT?

Thanks

File: clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp [hunk 4]
```diff
@@ -385,13 +387,15 @@ Error runSYCLLink(ArrayRef<std::string> Files, const ArgList &Args) {
...
     }
+
     OffloadingImage TheImage{};
     TheImage.TheImageKind = IMG_Object;
     TheImage.TheOffloadKind = OFK_SYCL;
     TheImage.StringData["triple"] =
         Args.MakeArgString(Args.getLastArgValue(OPT_triple_EQ));
     TheImage.StringData["arch"] =
         Args.MakeArgString(Args.getLastArgValue(OPT_arch_EQ));
+    TheImage.StringData["sycl_properties"] = Properties.writeJSON();
```
Comment (maksimsab) [inline clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp]: I would suggest simple "properties".

File: llvm/include/llvm/Frontend/Offloading/Utility.h [hunk 5]
```diff
@@ -159,6 +160,107 @@ namespace intel {
...
+/// A property set is a collection of PropertyValues, each identified by a name.
+/// A property set registry is a collection of property sets, each identified by
+/// a category name. The registry allows for the addition, removal, and
+/// retrieval of property sets and their properties.
+class PropertySetRegistry {
+  using PropertyMapTy = StringMap<unsigned>;
+  /// A property set. Preserves insertion order when iterating elements.
+  using PropertySet = MapVector<SmallString<16>, PropertyValue, PropertyMapTy>;
+  using MapTy = MapVector<SmallString<16>, PropertySet, PropertyMapTy>;
```
Comment (maksimsab) [inline llvm/include/llvm/Frontend/Offloading/Utility.h]: I would discourage usage of `SmallString<16>`.
https://github.com/intel/llvm/blob/sycl/sycl/doc/design/PropertySets.md
Typically, keys are larger than 16 and we wouldn't make use of this limit.
Also, LLVM's Programmers Manual discourages putting SmallString `as the member of a frequently-allocated heap data structure`.

I would rather make it `std::map<std::string, std::map<std::string, PropertyValue>>`.

We aren't going to have many keys here and it would make the interface extremely simple.

File: llvm/include/llvm/Frontend/Offloading/Utility.h [hunk 6]
```diff
@@ -159,6 +160,107 @@ namespace intel {
 /// the Intel runtime offload plugin.
 Error containerizeOpenMPSPIRVImage(std::unique_ptr<MemoryBuffer> &Binary);
 } // namespace intel
+
+namespace sycl {
+class PropertySetRegistry;
+
+// A property value. It can be either a 32-bit unsigned integer or a byte array.
+class PropertyValue {
```
Comment (maksimsab) [inline llvm/include/llvm/Frontend/Offloading/Utility.h]: I think that this class was overengineered from the very beginning.
This class can be replaced by simple `std::variant<uint64_t, std::string>`. In the text format it is not going to have much difference either it is uint64_t or uint32_t.

Do you have any objections againts the following declaration?
```cpp
using PropertyValue = std::variant<uint64_t, std::string>;
```

File: llvm/include/llvm/Frontend/Offloading/Utility.h [hunk 7]
```diff
@@ -159,6 +160,107 @@ namespace intel {
...
+
+  // Note: each enumeration value corresponds one-to-one to the
+  // index of the std::variant.
+  enum Type { PV_None = 0, PV_UInt32 = 1, PV_ByteArray = 2 };
+  Type getType() const { return static_cast<Type>(Value.index()); }
+
+  bool operator==(const PropertyValue &Rhs) const { return Value == Rhs.Value; }
+
+private:
+  std::variant<std::monostate, std::uint32_t, ByteArrayTy> Value = {};
```
Comment (maksimsab) [inline llvm/include/llvm/Frontend/Offloading/Utility.h]: I've just read what `monostate` is. It allows `std::variant` to be empty. However, I don't think we should allow our variants to be empty. Therefore, I would discourage a use of `monostate`.

**Comment by maksimsab** [comment]
Summary of my opinion.

I think, that the initial design was very poor and it doesn't worth to save it anyhow.

We could come up with the simple design like the following:
```cpp

namespace sycl {

using PropertyValue = std::variant<uint64_t, std::string>;

using Properties = std::map<std::string, std::map<std::string, PropertyValue>>;

void writePropertiesToJSON(const Properties &P, raw_ostream &O);

Expected<Properties> readPropertiesFromJSON(MemoryBufferRef MBR);

} // namespace sycl
```

Everything else could be thrown away. And this has the most chances to get upstreamed in my opinion.
Please, share your opinion.

File: llvm/lib/Frontend/Offloading/Utility.cpp [hunk 8]
```diff
@@ -459,3 +459,77 @@ Error offloading::intel::containerizeOpenMPSPIRVImage(
   Img = MemoryBuffer::getMemBufferCopy(YamlFile);
   return Error::success();
 }
+
+namespace llvm::offloading::sycl {
+
+void writePropertiesToJSON(const PropertySetRegistry &PSRegistry, raw_ostream &Out) {
+  json::OStream J(Out);
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/Utility.cpp]: Perhaps, we should include missing json header.

File: llvm/lib/Frontend/Offloading/Utility.cpp [hunk 9]
```diff
@@ -459,3 +459,77 @@ Error offloading::intel::containerizeOpenMPSPIRVImage(
...
+Expected<PropertySetRegistry>
+readPropertiesFromJSON(MemoryBufferRef Buf) {
+  PropertySetRegistry Res;
+  Expected<json::Value> V = json::parse(Buf.getBuffer());
+  if (!V)
+    return V.takeError();
+
+  const json::Object *O = V->getAsObject();
+  if (!O)
+    return createStringError("expected JSON object");
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/Utility.cpp]: I think we could communicate here something like: 
```cpp
  return createStringError("fail to obtain JSON Object from source: \"{0}\"", Buf.getIdentifier());
```

In the similar following errors we could also communicate some additional information that would help to track down the errors. Like, associated property names.

File: llvm/unittests/Frontend/PropertySetRegistryTest.cpp [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+      ASSERT_EQ(PropValue1, PropValue2);
+    }
+  }
+}
+
+TEST(PropertySetRegistryTest, PropertySetRegistry) {
+  PropertySetRegistry PSR;
+  PSR["Category1"]["Prop1"] = 42U;
+  PSR["Category1"]["Prop2"] = ByteArray(StringRef("Hello").bytes());
+  PSR["Category2"]["A"] = ByteArray{4, 16, 32};
```
Comment (maksimsab) [inline llvm/unittests/Frontend/PropertySetRegistryTest.cpp]: It could worth to add corner cases, like 0 and 255. Empty array as well.

File: llvm/unittests/Frontend/PropertySetRegistryTest.cpp [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  SmallString<0> Serialized;
+  raw_svector_ostream OS(Serialized);
+  writePropertiesToJSON(PSR, OS);
+  auto PSR2 = readPropertiesFromJSON({Serialized, ""});
+  if (auto Err = PSR2.takeError())
+    FAIL();
```
Comment (maksimsab) [inline llvm/unittests/Frontend/PropertySetRegistryTest.cpp]: It worth to propagate the error message further.

We could also add corner cases that tests error handling like arrays of integer that exceed 255 or negative integers.

File: llvm/lib/Frontend/Offloading/Utility.cpp [hunk 12]
```diff
@@ -459,3 +459,77 @@ Error offloading::intel::containerizeOpenMPSPIRVImage(
...
+          Vec.push_back(static_cast<unsigned char>(*Byte));
+        }
+        Prop = PropertyValue(std::move(Vec));
+      } else {
+        return createStringError("unsupported property type");
+      }
+
+      auto [It, Inserted] = PropSet.try_emplace(PropName.str(), std::move(Prop));
+      if (!Inserted)
+        return createStringError("duplicate property name");
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/Utility.cpp]: Maybe, this case is not possible in case of the correct JSON input.

File: llvm/lib/Frontend/Offloading/Utility.cpp [hunk 13]
```diff
@@ -459,3 +459,77 @@ Error offloading::intel::containerizeOpenMPSPIRVImage(
...
+    const json::Object *PropSetVal = Value.getAsObject();
+    if (!PropSetVal)
+      return createStringError("expected JSON array for properties");
+
+    PropertySet &PropSet = Res[CategoryName.str()];
+    for (const auto &[PropName, PropValueVal] : *PropSetVal) {
+      PropertyValue Prop;
+      if (std::optional<uint64_t> Val = PropValueVal.getAsUINT64()) {
+        Prop = PropertyValue(static_cast<uint32_t>(*Val));
+      } else if (const json::Array *Val = PropValueVal.getAsArray()) {
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/Utility.cpp]: Maybe, we could use a HEX-encoded string for the case of ByteArray. It would save some space in encoded JSON object because uint64_t is used for each byte and this is a 8x overhead.

**Review by maksimsab — CHANGES_REQUESTED** [review]
Great work in overall!

I have some requests about error messages.

File: llvm/include/llvm/Frontend/Offloading/PropertySet.h [hunk 14]
```diff
@@ -0,0 +1,22 @@
+#include "llvm/ADT/SmallVector.h"
```
Comment (maksimsab) [inline llvm/include/llvm/Frontend/Offloading/PropertySet.h:12]: Missing copyright header.

File: llvm/include/llvm/Frontend/Offloading/Utility.h
Comment (maksimsab) [reply llvm/include/llvm/Frontend/Offloading/Utility.h:1]: So where are we leaving it? I can see that they are currently duplicated in `PropertySet.h` and `Utility.h`.

File: llvm/lib/Frontend/Offloading/PropertySet.cpp [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+          default:
+            llvm_unreachable("unsupported property type");
+          }
+        }
+      });
+    }
+  });
+}
+
+template <typename... Ts> auto createStringErrorV(Ts &&...Args) {
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/PropertySet.cpp:40]: There is a corresponding overloading that you can use instead.

https://github.com/llvm/llvm-project/blob/0870c8838ba20edc0fb7efe60a0ee301bf01d209/llvm/include/llvm/Support/Error.h#L1328

File: llvm/lib/Frontend/Offloading/PropertySet.cpp [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+  const json::Object *O = V->getAsObject();
+  if (!O)
+    return createStringErrorV(
+        "error while deserializing property set registry: "
+        "expected JSON object, got {0}",
+        *V);
+
+  for (const auto &[CategoryName, Value] : *O) {
+    const json::Object *PropSetVal = Value.getAsObject();
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/PropertySet.cpp:77]: `Value.getAsArray()`?

File: llvm/lib/Frontend/Offloading/Utility.cpp (hunk 13 above)
Comment (maksimsab) [reply llvm/lib/Frontend/Offloading/Utility.cpp]: Any opinion on this one?
This is also possible to reuse base64 encoding that is available in LLVM.

File: llvm/lib/Frontend/Offloading/PropertySet.cpp [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+        }
+        Prop = PropertyValue(std::move(Vec));
+      } else {
+        return PropertyParseError("expected a uint64 or an array, got {0}",
+                                  PropValueVal);
+      }
+
+      auto [It, Inserted] =
+          PropSet.try_emplace(PropName.str(), std::move(Prop));
+      assert(Inserted);
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/PropertySet.cpp]: Please, leave some message that you had initially.

File: llvm/lib/Frontend/Offloading/PropertySet.cpp [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    const json::Object *PropSetVal = Value.getAsObject();
+    if (!PropSetVal)
+      return createStringErrorV("error while deserializing property set {0}: "
+                                "expected JSON array, got {1}",
+                                CategoryName.str(), Value);
+
+    PropertySet &PropSet = Res[CategoryName.str()];
+    for (const auto &[PropName, PropValueVal] : *PropSetVal) {
+      PropertyValue Prop;
+      auto PropertyParseError = [&](auto &&...MsgArgs) {
```
Comment (maksimsab) [inline llvm/lib/Frontend/Offloading/PropertySet.cpp]: This takes too much effort to read the loop body with this lambda.
If you extract the loop body in a separate function than it would not be so diffucult. Like the following:
```cpp
Expected<PropertyValue> retrievePropertyValueFromJSONObject(const json::Object *O);
```

Then you could stack errors in a simple way.

**Review by maksimsab — CHANGES_REQUESTED** [review]

File: llvm/lib/Frontend/Offloading/PropertySet.cpp (hunk 16 above)
Comment (jzc) [reply llvm/lib/Frontend/Offloading/PropertySet.cpp:77]: Good catch, looks like the error message is wrong, the code is right

File: llvm/lib/Frontend/Offloading/PropertySet.cpp (hunk 15 above)
Comment (jzc) [reply llvm/lib/Frontend/Offloading/PropertySet.cpp:40]: That uses `llvm::format`, which does not work with `json::Value`. We need to use `llvm::formatv`, which is why I added this function

File: llvm/include/llvm/Frontend/Offloading/Utility.h
Comment (jzc) [reply llvm/include/llvm/Frontend/Offloading/Utility.h:1]: Good catch, forgot to delete this

File: llvm/lib/Frontend/Offloading/Utility.cpp (hunk 13 above)
Comment (jzc) [reply llvm/lib/Frontend/Offloading/Utility.cpp]: True, I don't mind either way so I switched it to base64

**Review by maksimsab — APPROVED** [review]
LGTM.

File: llvm/lib/Frontend/Offloading/PropertySet.cpp [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+#include "llvm/Frontend/Offloading/PropertySet.h"
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/Support/MemoryBufferRef.h"
+
+using namespace llvm;
+using namespace llvm::offloading;
+
+void llvm::offloading::writePropertiesToJSON(
```
Comment (asudarsa) [inline llvm/lib/Frontend/Offloading/PropertySet.cpp:17]: Can we not do this?
```suggestion
void writePropertiesToJSON(
```

Comment (asudarsa) [reply llvm/lib/Frontend/Offloading/PropertySet.cpp:17]: If we can, pls update other places as well...Thanks

File: llvm/lib/Frontend/Offloading/CMakeLists.txt [hunk 20]
```diff
@@ -1,6 +1,7 @@
 add_llvm_component_library(LLVMFrontendOffloading
   Utility.cpp
   OffloadWrapper.cpp
+  PropertySet.cpp
```
Comment (asudarsa) [inline llvm/lib/Frontend/Offloading/CMakeLists.txt:4]: Please check shared build as well. It is a good idea to do that whenever you are introducing new files (or new headers in files)

Thanks

File: llvm/lib/Frontend/Offloading/PropertySet.cpp (hunk 19 above)
Comment (jzc) [reply llvm/lib/Frontend/Offloading/PropertySet.cpp:17]: It actually seems to be preferred: https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

File: llvm/include/llvm/Frontend/Offloading/PropertySet.h [hunk 21]
```diff
@@ -0,0 +1,29 @@
+///===- llvm/Frontend/Offloading/PropertySet.h ----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include "llvm/ADT/SmallVector.h"
```
Comment (asudarsa) [inline llvm/include/llvm/Frontend/Offloading/PropertySet.h:12]: Can we please add a small description?

//===----------------------------------------------------------------------===//
/// \file
///
/// This file defines PropertySetRegistry and PropertyValue types and provides
/// helper functions to translate PropertySetRegistry from/to JSON.
///
//===----------------------------------------------------------------------===//

**Review by asudarsa — APPROVED** [review]
LGTM. Think we can submit this upstream once we have a good PR description with an example and also details on where this will be used
Thanks
